### PR TITLE
Refactor/improve library build

### LINKED ISSUE #12: reduce library size
I've read articles about Rollup and how it creates smaller bundles than Webpack. I'd like to give it a try and check out what would be the differences comparing to the current approach.

### PR DESCRIPTION
**Pull request by AndyOGo**
fixes #12 

## Changes

- [x] create a library build
   - providing an ESM `module` version
   - providing an ES5 `main` version
   - provinding an UMD `unpgk` version
   - use [`microbundle`](https://github.com/developit/microbundle)
   ```sh
   Build "reaptcha" to dist:
       1275 B: index.js.gz
       1087 B: index.js.br
       1249 B: index.esm.js.gz
       1065 B: index.esm.js.br
       1319 B: index.umd.js.gz
       1118 B: index.umd.js.br
   ```

- [x] Upgrade outdated dependencies
   ```sh
   Package                                       Current   Wanted   Latest  Location
   @babel/core                                     7.5.5    7.8.6    7.8.6  reaptcha-monorepo
   @babel/plugin-proposal-class-properties         7.5.5    7.8.3    7.8.3  reaptcha-monorepo
   @babel/preset-env                               7.5.5    7.8.6    7.8.6  reaptcha-monorepo
   @babel/preset-flow                              7.0.0    7.8.3    7.8.3  reaptcha-monorepo
   @babel/preset-react                             7.0.0    7.8.3    7.8.3  reaptcha-monorepo
   @babel/register                                 7.5.5    7.8.6    7.8.6  reaptcha-monorepo
   @commitlint/cli                                 8.1.0    8.3.5    8.3.5  reaptcha-monorepo
   @commitlint/config-conventional                 8.1.0    8.3.4    8.3.4  reaptcha-monorepo
   ava                                      1.0.0-beta.8    1.4.1    3.4.0  reaptcha-monorepo
   babel-eslint                                   10.0.2   10.1.0   10.1.0  reaptcha-monorepo
   clean-webpack-plugin                            1.0.0    1.0.1    3.0.0  reaptcha-monorepo
   coveralls                                       3.0.2    3.0.9    3.0.9  reaptcha-monorepo
   cz-conventional-changelog                       3.0.2    3.1.0    3.1.0  reaptcha-monorepo
   enzyme                                          3.9.0    3.9.0   3.11.0  reaptcha-monorepo
   enzyme-adapter-react-16                        1.11.2   1.11.2   1.15.2  reaptcha-monorepo
   eslint                                          5.4.0   5.16.0    6.8.0  reaptcha-monorepo
   eslint-plugin-flowtype                          3.6.0    3.6.0    4.6.0  reaptcha-monorepo
   eslint-plugin-react                            7.11.1   7.18.3   7.18.3  reaptcha-monorepo
   flow-bin                                       0.96.0   0.96.0  0.119.1  reaptcha-monorepo
   husky                                           1.3.0    1.3.0    4.2.3  reaptcha-monorepo
   jsdom                                          15.0.0   15.0.0   16.2.0  reaptcha-monorepo
   lerna                                           3.2.1   3.20.2   3.20.2  reaptcha-monorepo
   nyc                                            13.2.0   13.3.0   15.0.0  reaptcha-monorepo
   prettier                                       1.14.2   1.19.1   1.19.1  reaptcha-monorepo
   pretty-quick                                    1.6.0   1.11.1    2.0.1  reaptcha-monorepo
   react-hot-loader                                4.3.5  4.12.19  4.12.19  reaptcha-monorepo
   sinon                                           7.2.4    7.5.0    9.0.0  reaptcha-monorepo
   uglifyjs-webpack-plugin                         2.0.0    2.2.0    2.2.0  reaptcha-monorepo
   webpack                                        4.17.1   4.41.6   4.41.6  reaptcha-monorepo
   webpack-cli                                     3.1.0   3.3.11   3.3.11  reaptcha-monorepo
   webpack-dev-server                              3.1.7   3.10.3   3.10.3  reaptcha-monorepo
   webpack-merge                                   4.1.4    4.2.2    4.2.2  reaptcha-monorepo
   ```

### CODE REVIEW

File: example/webpack.prod.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 const webpack = require('webpack');
 const merge = require('webpack-merge');
-const CleanWebpackPlugin = require('clean-webpack-plugin');
+const { CleanWebpackPlugin } = require('clean-webpack-plugin');
```
Comment (AndyOGo) [inline example/webpack.prod.js]: Usage has changed, and `dist` is default dir.
https://github.com/johnagan/clean-webpack-plugin#usage

File: lib/index.test.js [hunk 2]
```diff
@@ -440,6 +440,7 @@ test.serial('should throw error on double render', t => {
     t.true(renderSpy.calledOnce);
     return t.throwsAsync(
       instance.renderExplicitly(),
+      null,
```
Comment (AndyOGo) [inline lib/index.test.js]: Signature of [`thowsAsync`](https://github.com/avajs/ava/blob/master/docs/03-assertions.md#throwsasyncthrower-expectation-message) has changed

File: lib/package.json [hunk 3]
```diff
@@ -12,15 +12,20 @@
...
     "index.d.ts"
   ],
+  "source": "index.js",
   "main": "dist/index.js",
+  "module": "dist/index.mjs",
+  "unpkg": "dist/index.umd.js",
   "typings": "./index.d.ts",
   "scripts": {
-    "build": "webpack --config webpack.config.js"
+    "build": "microbundle --jsx React.createElement"
```
Comment (AndyOGo) [inline lib/package.json]:
```sh
ERROR in ../lib/dist/index.mjs 1:3393-3394
Can't import the named export 'Component' from non EcmaScript module (only default export is available)
 @ ./src/modules/example.js
 @ ./src/containers/index.js
 @ ./src/index.js
```
https://github.com/developit/microbundle/issues/426

File: lib/index.js [hunk 4]
```diff
@@ -1,6 +1,6 @@
 /* @flow */
 
-import React, { Component } from 'react';
+import React from 'react';
```
Comment (AndyOGo) [inline lib/index.js]: For `.mjs` files (which is generated by microbundle) webpack only allows strict CJS imports (with named exports).
Further reading: https://lifesaver.codes/answer/webpack-4-compatibility-issue

File: lib/package.json (hunk 3 above)
Comment (AndyOGo) [reply lib/package.json]: fixed by renaming **module** dist from `dist/index.mjs` to `dist/index.esm.js`.
Reason webpack treat's `*.mjs` modules very strictly in regard to named imports of commonJS modules.

File: lib/index.js (hunk 4 above)
Comment (AndyOGo) [reply lib/index.js]: Fixed by not using `*.mjs` file extension for **module** field, so this is reverted.

**Comment by sarneeh** [comment]
@AndyOGo Thank you for this PR! Amazing work! 😮 I need some time to review it, give me time to monday please, I'll try to catch up on the weekend 😃

**Comment by AndyOGo** [comment]
@sarneeh Your welcome!
Your implementation is the cleanest I found on the web, it helps us really. So if I can support you, Im happy to help.

In regard to maintaince of dependencies, you may be interested in https://dependabot.com/
It's free for github and takes a lot of the burden of version management for you.

For the example app and the demo site I haven't made any chances yet, besides of updating webpack.
Other options would be https://parceljs.org/

Take the time you need, Im looking forward to your feedback.

**Comment by AndyOGo** [comment]
Hey @sarneeh 
How are you doing?
Are you very busy with work?
Would be glad to hear from you.

Have a nice weekend.

**Comment by sarneeh** [comment]
Hey @AndyOGo, I'm sorry but I didn't have time to maintain this package and handle any PR's 😢  

I have a little bit more free time right now so I decided to rewrite the whole package in TypeScript and also improve the build at once. Because of this, I will close your PR and try to reuse the work that you did so that it won't be wasted. I'm sorry that your original code won't get into the codebase, I hope you won't get mad at me 😅  It's just a waste of time to merge and test this right now as I'm rewriting it from scratch anyways.

If you're interested in future work on this package, I can add you to the list of contributors! Let me know 😃

**Comment by sarneeh** [comment]
Changed my mind. Will actually try to merge in your PR 🤪 Gave this idea another thought and realised it's a waste of time to rewrite everything 😅

**Comment by AndyOGo** [comment]
close in favour of #235 235

**Pull request closed without merging**